File: Source/DafnyStandardLibraries/binaries/GoLang/translated-go/src/Std_Unicode_Utf16EncodingForm/Std_Unicode_Utf16EncodingForm.go
// Package Std_Unicode_Utf16EncodingForm
// Dafny module Std_Unicode_Utf16EncodingForm compiled into Go

package Std_Unicode_Utf16EncodingForm

import (
  _dafny "dafny"
  os "os"
  _System "System_"
  Std_Wrappers "Std_Wrappers"
  Std_BoundedInts "Std_BoundedInts"
  Std_Base64 "Std_Base64"
  Std_Relations "Std_Relations"
  Std_Math "Std_Math"
  Std_Collections_Seq "Std_Collections_Seq"
  Std_Collections_Array "Std_Collections_Array"
  Std_Collections_Imap "Std_Collections_Imap"
  Std_Functions "Std_Functions"
  Std_Collections_Iset "Std_Collections_Iset"
  Std_Collections_Map "Std_Collections_Map"
  Std_Collections_Set "Std_Collections_Set"
  Std_Collections "Std_Collections"
  Std_DynamicArray "Std_DynamicArray"
  Std_FileIO "Std_FileIO"
  Std_Arithmetic_GeneralInternals "Std_Arithmetic_GeneralInternals"
  Std_Arithmetic_MulInternalsNonlinear "Std_Arithmetic_MulInternalsNonlinear"
  Std_Arithmetic_MulInternals "Std_Arithmetic_MulInternals"
  Std_Arithmetic_Mul "Std_Arithmetic_Mul"
  Std_Arithmetic_ModInternalsNonlinear "Std_Arithmetic_ModInternalsNonlinear"
  Std_Arithmetic_DivInternalsNonlinear "Std_Arithmetic_DivInternalsNonlinear"
  Std_Arithmetic_ModInternals "Std_Arithmetic_ModInternals"
  Std_Arithmetic_DivInternals "Std_Arithmetic_DivInternals"
  Std_Arithmetic_DivMod "Std_Arithmetic_DivMod"
  Std_Arithmetic_Power "Std_Arithmetic_Power"
  Std_Arithmetic_Logarithm "Std_Arithmetic_Logarithm"
  Std_Arithmetic_Power2 "Std_Arithmetic_Power2"
  Std_Arithmetic "Std_Arithmetic"
  Std_Strings_HexConversion "Std_Strings_HexConversion"
  Std_Strings_DecimalConversion "Std_Strings_DecimalConversion"
  Std_Strings_CharStrEscaping "Std_Strings_CharStrEscaping"
  Std_Strings "Std_Strings"
  Std_Unicode_Base "Std_Unicode_Base"
  Std_Unicode_Utf8EncodingForm "Std_Unicode_Utf8EncodingForm"
)
var _ _dafny.Dummy__
var _ = os.Args
var _ _System.Dummy__
var _ Std_Wrappers.Dummy__
var _ Std_BoundedInts.Dummy__
var _ Std_Base64.Dummy__
var _ Std_Relations.Dummy__
var _ Std_Math.Dummy__
var _ Std_Collections_Seq.Dummy__
var _ Std_Collections_Array.Dummy__
var _ Std_Collections_Imap.Dummy__
var _ Std_Functions.Dummy__
var _ Std_Collections_Iset.Dummy__
var _ Std_Collections_Map.Dummy__
var _ Std_Collections_Set.Dummy__
var _ Std_Collections.Dummy__
var _ Std_DynamicArray.Dummy__
var _ Std_FileIO.Dummy__
var _ Std_Arithmetic_GeneralInternals.Dummy__
var _ Std_Arithmetic_MulInternalsNonlinear.Dummy__
var _ Std_Arithmetic_MulInternals.Dummy__
var _ Std_Arithmetic_Mul.Dummy__
var _ Std_Arithmetic_ModInternalsNonlinear.Dummy__
var _ Std_Arithmetic_DivInternalsNonlinear.Dummy__
var _ Std_Arithmetic_ModInternals.Dummy__
var _ Std_Arithmetic_DivInternals.Dummy__
var _ Std_Arithmetic_DivMod.Dummy__
var _ Std_Arithmetic_Power.Dummy__
var _ Std_Arithmetic_Logarithm.Dummy__
var _ Std_Arithmetic_Power2.Dummy__
var _ Std_Arithmetic.Dummy__
var _ Std_Strings_HexConversion.Dummy__
var _ Std_Strings_DecimalConversion.Dummy__
var _ Std_Strings_CharStrEscaping.Dummy__
var _ Std_Strings.Dummy__
var _ Std_Unicode_Base.Dummy__
var _ Std_Unicode_Utf8EncodingForm.Dummy__

type Dummy__ struct{}


// Definition of class Default__
type Default__ struct {
  dummy byte
}

func New_Default___() *Default__ {
  _this := Default__{}

  return &_this
}

type CompanionStruct_Default___ struct {
}
var Companion_Default___ = CompanionStruct_Default___ {
}

func (_this *Default__) Equals(other *Default__) bool {
  return _this == other
}

func (_this *Default__) EqualsGeneric(x interface{}) bool {
  other, ok := x.(*Default__)
  return ok && _this.Equals(other)
}

func (*Default__) String() string {
  return "Std_Unicode_Utf16EncodingForm.Default__"
}
func (_this *Default__) ParentTraits_() []*_dafny.TraitID {
  return [](*_dafny.TraitID){};
}
var _ _dafny.TraitOffspring = &Default__{}

func (_static *CompanionStruct_Default___) IsMinimalWellFormedCodeUnitSubsequence(s _dafny.Sequence) bool {
  if ((_dafny.IntOfUint32((s).Cardinality())).Cmp(_dafny.One) == 0) {
    return Companion_Default___.IsWellFormedSingleCodeUnitSequence(s)
  } else if ((_dafny.IntOfUint32((s).Cardinality())).Cmp(_dafny.IntOfInt64(2)) == 0) {
    var _231_b bool = Companion_Default___.IsWellFormedDoubleCodeUnitSequence(s)
    _ = _231_b
    return _231_b
  } else {
    return false
  }
}
func (_static *CompanionStruct_Default___) IsWellFormedSingleCodeUnitSequence(s _dafny.Sequence) bool {
  var _232_firstWord uint16 = (s).Select(0).(uint16)
  _ = _232_firstWord
  return (((uint16(0)) <= (_232_firstWord)) && ((_232_firstWord) <= (uint16(55295)))) || (((uint16(57344)) <= (_232_firstWord)) && ((_232_firstWord) <= (uint16(65535))))
}
func (_static *CompanionStruct_Default___) IsWellFormedDoubleCodeUnitSequence(s _dafny.Sequence) bool {
  var _233_firstWord uint16 = (s).Select(0).(uint16)
  _ = _233_firstWord
  var _234_secondWord uint16 = (s).Select(1).(uint16)
  _ = _234_secondWord
  return (((uint16(55296)) <= (_233_firstWord)) && ((_233_firstWord) <= (uint16(56319)))) && (((uint16(56320)) <= (_234_secondWord)) && ((_234_secondWord) <= (uint16(57343))))
}
func (_static *CompanionStruct_Default___) SplitPrefixMinimalWellFormedCodeUnitSubsequence(s _dafny.Sequence) Std_Wrappers.Option {
  if (((_dafny.IntOfUint32((s).Cardinality())).Cmp(_dafny.One) >= 0) && (Companion_Default___.IsWellFormedSingleCodeUnitSequence((s).Take(1)))) {
    return Std_Wrappers.Companion_Option_.Create_Some_((s).Take(1))
  } else if (((_dafny.IntOfUint32((s).Cardinality())).Cmp(_dafny.IntOfInt64(2)) >= 0) && (Companion_Default___.IsWellFormedDoubleCodeUnitSequence((s).Take(2)))) {
    return Std_Wrappers.Companion_Option_.Create_Some_((s).Take(2))
  } else {
    return Std_Wrappers.Companion_Option_.Create_None_()
  }
}
func (_static *CompanionStruct_Default___) EncodeScalarValue(v uint32) _dafny.Sequence {
  if ((((uint32(0)) <= (v)) && ((v) <= (uint32(55295)))) || (((uint32(57344)) <= (v)) && ((v) <= (uint32(65535))))) {
    return Companion_Default___.EncodeScalarValueSingleWord(v)
  } else {
    return Companion_Default___.EncodeScalarValueDoubleWord(v)
  }
}
func (_static *CompanionStruct_Default___) EncodeScalarValueSingleWord(v uint32) _dafny.Sequence {
  var _235_firstWord uint16 = uint16(v)
  _ = _235_firstWord
  return _dafny.SeqOf(_235_firstWord)
}
func (_static *CompanionStruct_Default___) EncodeScalarValueDoubleWord(v uint32) _dafny.Sequence {
  var _236_x2 uint16 = uint16((v) & (uint32(1023)))
  _ = _236_x2
  var _237_x1 uint8 = uint8(((v) & (uint32(64512))) >> (uint8(10)))
  _ = _237_x1
  var _238_u uint8 = uint8(((v) & (uint32(2031616))) >> (uint8(16)))
  _ = _238_u
  var _239_w uint8 = uint8((((_238_u) - (func () uint8 { return  (uint8(1)) })()) & 0x1F))
  _ = _239_w
  var _240_firstWord uint16 = ((uint16(55296)) | ((uint16(_239_w)) << (uint8(6)))) | (uint16(_237_x1))
  _ = _240_firstWord
  var _241_secondWord uint16 = (uint16(56320)) | (uint16(_236_x2))
  _ = _241_secondWord
  return _dafny.SeqOf(_240_firstWord, _241_secondWord)
}
func (_static *CompanionStruct_Default___) DecodeMinimalWellFormedCodeUnitSubsequence(m _dafny.Sequence) uint32 {
  if ((_dafny.IntOfUint32((m).Cardinality())).Cmp(_dafny.One) == 0) {
    return Companion_Default___.DecodeMinimalWellFormedCodeUnitSubsequenceSingleWord(m)
  } else {
    return Companion_Default___.DecodeMinimalWellFormedCodeUnitSubsequenceDoubleWord(m)
  }
}
func (_static *CompanionStruct_Default___) DecodeMinimalWellFormedCodeUnitSubsequenceSingleWord(m _dafny.Sequence) uint32 {
  var _242_firstWord uint16 = (m).Select(0).(uint16)
  _ = _242_firstWord
  var _243_x uint16 = (_242_firstWord)
  _ = _243_x
  return uint32(_243_x)
}
func (_static *CompanionStruct_Default___) DecodeMinimalWellFormedCodeUnitSubsequenceDoubleWord(m _dafny.Sequence) uint32 {
  var _244_firstWord uint16 = (m).Select(0).(uint16)
  _ = _244_firstWord
  var _245_secondWord uint16 = (m).Select(1).(uint16)
  _ = _245_secondWord
  var _246_x2 uint32 = uint32((_245_secondWord) & (uint16(1023)))
  _ = _246_x2
  var _247_x1 uint32 = uint32((_244_firstWord) & (uint16(63)))
  _ = _247_x1
  var _248_w uint32 = uint32(((_244_firstWord) & (uint16(960))) >> (uint8(6)))
  _ = _248_w
  var _249_u uint32 = ((((_248_w) + (uint32(1))) & 0xFFFFFF))
  _ = _249_u
  var _250_v uint32 = (((((_249_u) << (uint8(16))) & 0xFFFFFF)) | ((((_247_x1) << (uint8(10))) & 0xFFFFFF))) | ((_246_x2))
  _ = _250_v
  return _250_v
}
func (_static *CompanionStruct_Default___) PartitionCodeUnitSequenceChecked(s _dafny.Sequence) Std_Wrappers.Option {
  var maybeParts Std_Wrappers.Option = Std_Wrappers.Companion_Option_.Default()
  _ = maybeParts
  if (_dafny.Companion_Sequence_.Equal(s, _dafny.SeqOf())) {
    maybeParts = Std_Wrappers.Companion_Option_.Create_Some_(_dafny.SeqOf())
    return maybeParts
  }
  var _251_result _dafny.Sequence
  _ = _251_result
  _251_result = _dafny.SeqOf()
  var _252_rest _dafny.Sequence
  _ = _252_rest
  _252_rest = s
  for (_dafny.IntOfUint32((_252_rest).Cardinality())).Sign() == 1 {
    var _253_prefix _dafny.Sequence
    _ = _253_prefix
    var _254_valueOrError0 Std_Wrappers.Option = Std_Wrappers.Companion_Option_.Default()
    _ = _254_valueOrError0
    _254_valueOrError0 = Companion_Default___.SplitPrefixMinimalWellFormedCodeUnitSubsequence(_252_rest)
    if ((_254_valueOrError0).IsFailure()) {
      maybeParts = (_254_valueOrError0).PropagateFailure()
      return maybeParts
    }
    _253_prefix = (_254_valueOrError0).Extract().(_dafny.Sequence)
    _251_result = _dafny.Companion_Sequence_.Concatenate(_251_result, _dafny.SeqOf(_253_prefix))
    _252_rest = (_252_rest).Drop((_dafny.IntOfUint32((_253_prefix).Cardinality())).Uint32())
  }
  maybeParts = Std_Wrappers.Companion_Option_.Create_Some_(_251_result)
  return maybeParts
  return maybeParts
}
func (_static *CompanionStruct_Default___) PartitionCodeUnitSequence(s _dafny.Sequence) _dafny.Sequence {
  return (Companion_Default___.PartitionCodeUnitSequenceChecked(s)).Extract().(_dafny.Sequence)
}
func (_static *CompanionStruct_Default___) IsWellFormedCodeUnitSequence(s _dafny.Sequence) bool {
  return (Companion_Default___.PartitionCodeUnitSequenceChecked(s)).Is_Some()
}
func (_static *CompanionStruct_Default___) EncodeScalarSequence(vs _dafny.Sequence) _dafny.Sequence {
  var s _dafny.Sequence = Companion_WellFormedCodeUnitSeq_.Witness()
  _ = s
  s = _dafny.SeqOf()
  var _lo1 _dafny.Int = _dafny.Zero
  _ = _lo1
  for _255_i := _dafny.IntOfUint32((vs).Cardinality()); _lo1.Cmp(_255_i) < 0;  {
    _255_i = _255_i.Minus(_dafny.One)
    var _256_next _dafny.Sequence
    _ = _256_next
    _256_next = Companion_Default___.EncodeScalarValue((vs).Select((_255_i).Uint32()).(uint32))
    s = _dafny.Companion_Sequence_.Concatenate(_256_next, s)
  }
  return s
}
func (_static *CompanionStruct_Default___) DecodeCodeUnitSequence(s _dafny.Sequence) _dafny.Sequence {
  var _257_parts _dafny.Sequence = Companion_Default___.PartitionCodeUnitSequence(s)
  _ = _257_parts
  var _258_vs _dafny.Sequence = Std_Collections_Seq.Companion_Default___.Map(func (coer8 func (_dafny.Sequence) uint32) func (interface{}) interface{} {
    return func (arg10 interface{}) interface{} {
      return coer8(arg10.(_dafny.Sequence))
    }
  }(Companion_Default___.DecodeMinimalWellFormedCodeUnitSubsequence), _257_parts)
  _ = _258_vs
  return _258_vs
}
func (_static *CompanionStruct_Default___) DecodeCodeUnitSequenceChecked(s _dafny.Sequence) Std_Wrappers.Option {
  var maybeVs Std_Wrappers.Option = Std_Wrappers.Companion_Option_.Default()
  _ = maybeVs
  var _259_maybeParts Std_Wrappers.Option
  _ = _259_maybeParts
  _259_maybeParts = Companion_Default___.PartitionCodeUnitSequenceChecked(s)
  if ((_259_maybeParts).Is_None()) {
    maybeVs = Std_Wrappers.Companion_Option_.Create_None_()
    return maybeVs
  }
  var _260_parts _dafny.Sequence
  _ = _260_parts
  _260_parts = (_259_maybeParts).Dtor_value().(_dafny.Sequence)
  var _261_vs _dafny.Sequence
  _ = _261_vs
  _261_vs = Std_Collections_Seq.Companion_Default___.Map(func (coer9 func (_dafny.Sequence) uint32) func (interface{}) interface{} {
    return func (arg11 interface{}) interface{} {
      return coer9(arg11.(_dafny.Sequence))
    }
  }(Companion_Default___.DecodeMinimalWellFormedCodeUnitSubsequence), _260_parts)
  maybeVs = Std_Wrappers.Companion_Option_.Create_Some_(_261_vs)
  return maybeVs
  return maybeVs
}
// End of class Default__

// Definition of class WellFormedCodeUnitSeq
type WellFormedCodeUnitSeq struct {
}

func New_WellFormedCodeUnitSeq_() *WellFormedCodeUnitSeq {
  _this := WellFormedCodeUnitSeq{}

  return &_this
}

type CompanionStruct_WellFormedCodeUnitSeq_ struct {
}
var Companion_WellFormedCodeUnitSeq_ = CompanionStruct_WellFormedCodeUnitSeq_ {
}

func (*WellFormedCodeUnitSeq) String() string {
  return "Std_Unicode_Utf16EncodingForm.WellFormedCodeUnitSeq"
}
func (_this *CompanionStruct_WellFormedCodeUnitSeq_) Witness() _dafny.Sequence {
  return _dafny.SeqOf()
}
// End of class WellFormedCodeUnitSeq

func Type_WellFormedCodeUnitSeq_() _dafny.TypeDescriptor {
  return type_WellFormedCodeUnitSeq_{}
}

type type_WellFormedCodeUnitSeq_ struct {
}

func (_this type_WellFormedCodeUnitSeq_) Default() interface{} {
  return Companion_WellFormedCodeUnitSeq_.Witness()
}

func (_this type_WellFormedCodeUnitSeq_) String() string {
  return "Std_Unicode_Utf16EncodingForm.WellFormedCodeUnitSeq"
}

// Definition of class MinimalWellFormedCodeUnitSeq
type MinimalWellFormedCodeUnitSeq struct {
}

func New_MinimalWellFormedCodeUnitSeq_() *MinimalWellFormedCodeUnitSeq {
  _this := MinimalWellFormedCodeUnitSeq{}

  return &_this
}

type CompanionStruct_MinimalWellFormedCodeUnitSeq_ struct {
}
var Companion_MinimalWellFormedCodeUnitSeq_ = CompanionStruct_MinimalWellFormedCodeUnitSeq_ {
}

func (*MinimalWellFormedCodeUnitSeq) String() string {
  return "Std_Unicode_Utf16EncodingForm.MinimalWellFormedCodeUnitSeq"
}
// End of class MinimalWellFormedCodeUnitSeq

func Type_MinimalWellFormedCodeUnitSeq_() _dafny.TypeDescriptor {
  return type_MinimalWellFormedCodeUnitSeq_{}
}

type type_MinimalWellFormedCodeUnitSeq_ struct {
}

func (_this type_MinimalWellFormedCodeUnitSeq_) Default() interface{} {
  return _dafny.EmptySeq
}

func (_this type_MinimalWellFormedCodeUnitSeq_) String() string {
  return "Std_Unicode_Utf16EncodingForm.MinimalWellFormedCodeUnitSeq"
}

File: Source/DafnyStandardLibraries/binaries/Python/translated-py/Std_Unicode_Utf16EncodingForm.py
import sys
from typing import Callable, Any, TypeVar, NamedTuple
from math import floor
from itertools import count

import module_
import _dafny
import System_
import Std_Wrappers
import Std_BoundedInts
import Std_Base64
import Std_Relations
import Std_Math
import Std_Collections_Seq
import Std_Collections_Array
import Std_Collections_Imap
import Std_Functions
import Std_Collections_Iset
import Std_Collections_Map
import Std_Collections_Set
import Std_Collections
import Std_DynamicArray
import Std_FileIO
import Std_Arithmetic_GeneralInternals
import Std_Arithmetic_MulInternalsNonlinear
import Std_Arithmetic_MulInternals
import Std_Arithmetic_Mul
import Std_Arithmetic_ModInternalsNonlinear
import Std_Arithmetic_DivInternalsNonlinear
import Std_Arithmetic_ModInternals
import Std_Arithmetic_DivInternals
import Std_Arithmetic_DivMod
import Std_Arithmetic_Power
import Std_Arithmetic_Logarithm
import Std_Arithmetic_Power2
import Std_Arithmetic
import Std_Strings_HexConversion
import Std_Strings_DecimalConversion
import Std_Strings_CharStrEscaping
import Std_Strings
import Std_Unicode_Base
import Std_Unicode_Utf8EncodingForm

# Module: Std_Unicode_Utf16EncodingForm

class default__:
    def  __init__(self):
        pass

    @staticmethod
    def IsMinimalWellFormedCodeUnitSubsequence(s):
        if (len(s)) == (1):
            return default__.IsWellFormedSingleCodeUnitSequence(s)
        elif (len(s)) == (2):
            d_227_b_ = default__.IsWellFormedDoubleCodeUnitSequence(s)
            return d_227_b_
        elif True:
            return False

    @staticmethod
    def IsWellFormedSingleCodeUnitSequence(s):
        d_228_firstWord_ = (s)[0]
        return (((0) <= (d_228_firstWord_)) and ((d_228_firstWord_) <= (55295))) or (((57344) <= (d_228_firstWord_)) and ((d_228_firstWord_) <= (65535)))

    @staticmethod
    def IsWellFormedDoubleCodeUnitSequence(s):
        d_229_firstWord_ = (s)[0]
        d_230_secondWord_ = (s)[1]
        return (((55296) <= (d_229_firstWord_)) and ((d_229_firstWord_) <= (56319))) and (((56320) <= (d_230_secondWord_)) and ((d_230_secondWord_) <= (57343)))

    @staticmethod
    def SplitPrefixMinimalWellFormedCodeUnitSubsequence(s):
        if ((len(s)) >= (1)) and (default__.IsWellFormedSingleCodeUnitSequence(_dafny.SeqWithoutIsStrInference((s)[:1:]))):
            return Std_Wrappers.Option_Some(_dafny.SeqWithoutIsStrInference((s)[:1:]))
        elif ((len(s)) >= (2)) and (default__.IsWellFormedDoubleCodeUnitSequence(_dafny.SeqWithoutIsStrInference((s)[:2:]))):
            return Std_Wrappers.Option_Some(_dafny.SeqWithoutIsStrInference((s)[:2:]))
        elif True:
            return Std_Wrappers.Option_None()

    @staticmethod
    def EncodeScalarValue(v):
        if (((0) <= (v)) and ((v) <= (55295))) or (((57344) <= (v)) and ((v) <= (65535))):
            return default__.EncodeScalarValueSingleWord(v)
        elif True:
            return default__.EncodeScalarValueDoubleWord(v)

    @staticmethod
    def EncodeScalarValueSingleWord(v):
        d_231_firstWord_ = v
        return _dafny.SeqWithoutIsStrInference([d_231_firstWord_])

    @staticmethod
    def EncodeScalarValueDoubleWord(v):
        d_232_x2_ = (v) & (1023)
        d_233_x1_ = ((v) & (64512)) >> (10)
        d_234_u_ = ((v) & (2031616)) >> (16)
        d_235_w_ = ((d_234_u_) - (1)) & ((1 << 5) - 1)
        d_236_firstWord_ = ((55296) | (((d_235_w_) << (6)) & ((1 << 16) - 1))) | (d_233_x1_)
        d_237_secondWord_ = (56320) | (d_232_x2_)
        return _dafny.SeqWithoutIsStrInference([d_236_firstWord_, d_237_secondWord_])

    @staticmethod
    def DecodeMinimalWellFormedCodeUnitSubsequence(m):
        if (len(m)) == (1):
            return default__.DecodeMinimalWellFormedCodeUnitSubsequenceSingleWord(m)
        elif True:
            return default__.DecodeMinimalWellFormedCodeUnitSubsequenceDoubleWord(m)

    @staticmethod
    def DecodeMinimalWellFormedCodeUnitSubsequenceSingleWord(m):
        d_238_firstWord_ = (m)[0]
        d_239_x_ = d_238_firstWord_
        return d_239_x_

    @staticmethod
    def DecodeMinimalWellFormedCodeUnitSubsequenceDoubleWord(m):
        d_240_firstWord_ = (m)[0]
        d_241_secondWord_ = (m)[1]
        d_242_x2_ = (d_241_secondWord_) & (1023)
        d_243_x1_ = (d_240_firstWord_) & (63)
        d_244_w_ = ((d_240_firstWord_) & (960)) >> (6)
        d_245_u_ = ((d_244_w_) + (1)) & ((1 << 24) - 1)
        d_246_v_ = ((((d_245_u_) << (16)) & ((1 << 24) - 1)) | (((d_243_x1_) << (10)) & ((1 << 24) - 1))) | (d_242_x2_)
        return d_246_v_

    @staticmethod
    def PartitionCodeUnitSequenceChecked(s):
        maybeParts: Std_Wrappers.Option = Std_Wrappers.Option.default()()
        if (s) == (_dafny.SeqWithoutIsStrInference([])):
            maybeParts = Std_Wrappers.Option_Some(_dafny.SeqWithoutIsStrInference([]))
            return maybeParts
        d_247_result_: _dafny.Seq
        d_247_result_ = _dafny.SeqWithoutIsStrInference([])
        d_248_rest_: _dafny.Seq
        d_248_rest_ = s
        while (len(d_248_rest_)) > (0):
            d_249_prefix_: _dafny.Seq
            d_250_valueOrError0_: Std_Wrappers.Option = Std_Wrappers.Option.default()()
            d_250_valueOrError0_ = default__.SplitPrefixMinimalWellFormedCodeUnitSubsequence(d_248_rest_)
            if (d_250_valueOrError0_).IsFailure():
                maybeParts = (d_250_valueOrError0_).PropagateFailure()
                return maybeParts
            d_249_prefix_ = (d_250_valueOrError0_).Extract()
            d_247_result_ = (d_247_result_) + (_dafny.SeqWithoutIsStrInference([d_249_prefix_]))
            d_248_rest_ = _dafny.SeqWithoutIsStrInference((d_248_rest_)[len(d_249_prefix_)::])
        maybeParts = Std_Wrappers.Option_Some(d_247_result_)
        return maybeParts
        return maybeParts

    @staticmethod
    def PartitionCodeUnitSequence(s):
        return (default__.PartitionCodeUnitSequenceChecked(s)).Extract()

    @staticmethod
    def IsWellFormedCodeUnitSequence(s):
        return (default__.PartitionCodeUnitSequenceChecked(s)).is_Some

    @staticmethod
    def EncodeScalarSequence(vs):
        s: _dafny.Seq = WellFormedCodeUnitSeq.default()
        s = _dafny.SeqWithoutIsStrInference([])
        lo1_ = 0
        for d_251_i_ in range(len(vs)-1, lo1_-1, -1):
            d_252_next_: _dafny.Seq
            d_252_next_ = default__.EncodeScalarValue((vs)[d_251_i_])
            s = (d_252_next_) + (s)
        return s

    @staticmethod
    def DecodeCodeUnitSequence(s):
        d_253_parts_ = default__.PartitionCodeUnitSequence(s)
        d_254_vs_ = Std_Collections_Seq.default__.Map(default__.DecodeMinimalWellFormedCodeUnitSubsequence, d_253_parts_)
        return d_254_vs_

    @staticmethod
    def DecodeCodeUnitSequenceChecked(s):
        maybeVs: Std_Wrappers.Option = Std_Wrappers.Option.default()()
        d_255_maybeParts_: Std_Wrappers.Option
        d_255_maybeParts_ = default__.PartitionCodeUnitSequenceChecked(s)
        if (d_255_maybeParts_).is_None:
            maybeVs = Std_Wrappers.Option_None()
            return maybeVs
        d_256_parts_: _dafny.Seq
        d_256_parts_ = (d_255_maybeParts_).value
        d_257_vs_: _dafny.Seq
        d_257_vs_ = Std_Collections_Seq.default__.Map(default__.DecodeMinimalWellFormedCodeUnitSubsequence, d_256_parts_)
        maybeVs = Std_Wrappers.Option_Some(d_257_vs_)
        return maybeVs
        return maybeVs


class WellFormedCodeUnitSeq:
    def  __init__(self):
        pass

    @staticmethod
    def default():
        return _dafny.SeqWithoutIsStrInference([])

class MinimalWellFormedCodeUnitSeq:
    def  __init__(self):
        pass

    @staticmethod
    def default():
        return _dafny.Seq({})
